sklearn/cluster/_hierarchical.pyx

```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
         raise ValueError("Input MST array must be sorted by weight")
 
     return _single_linkage_label(L)
+  
+
+# Implements MST-LINKAGE-CORE from https://arxiv.org/abs/1109.2378
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] mst_linkage_core(
+        np.ndarray[DTYPE_t, ndim=2, mode='c'] raw_data,
+        DistanceMetric dist_metric):
+    """
+    Compute the the necessary elements of a minimum spanning
+    tree for computation of single linkage clustering. This
+    represents the MST-LINKAGE-CORE algorithm (Figure 6) from
+    *Modern hierarchical, agglomerative clustering algorithms*
+    by Daniel Mullner (https://arxiv.org/abs/1109.2378).
+
+    In contrast to the scipy implementation is never computes
+    a full distance matrix, generating distances only as they
+    are needed and releasing them when no longer needed.
+
+    Parameters
+    ----------
+    raw_data: array of shape (n_samples, n_features)
+        The array of feature data to be clustered.
+
+    dist_metric: DistanceMetric
+        A distance metric function (from sklearn.neighbors.dist_metrics)
+        used to compute distances.
+
+    Returns
+    -------
+    mst_core_data: array of shape (n_samples, 3)
+        An array providing information from which one
+        can either compute an MST, or the linkage hierarchy
+        very efficiently. See https://arxiv.org/abs/1109.2378
+        algorithm MST-LINKAGE-CORE for more details.
+    """
+
+    cdef DTYPE_t[:] current_distances
+    cdef DTYPE_t * raw_data_ptr
+    cdef np.int8_t[:] in_tree
+    cdef DTYPE_t[:, ::1] raw_data_view
+    cdef DTYPE_t[:, ::1] result
+
+    cdef np.ndarray label_filter
+
+    cdef ITYPE_t current_node
+    cdef ITYPE_t new_node
+    cdef ITYPE_t i
+    cdef ITYPE_t j
+    cdef ITYPE_t dim
+    cdef ITYPE_t num_features
+
+    cdef DTYPE_t right_value
+    cdef DTYPE_t left_value
+    cdef DTYPE_t new_distance
+
+    dim = raw_data.shape[0]
+    num_features = raw_data.shape[1]
+
+    raw_data_view = (<DTYPE_t[:raw_data.shape[0], :raw_data.shape[1]:1]> (
+        <DTYPE_t *> raw_data.data))
+    raw_data_ptr = (<DTYPE_t *> &raw_data_view[0, 0])
+
+    result = np.zeros((dim - 1, 3))
+    in_tree = np.zeros(dim, dtype=np.int8)
+    current_node = 0
+    current_distances = np.infty * np.ones(dim)
```

Comment:
`np.full` is preferred for this

Also we typically use ` from numpy.math cimport INFINITY`  for representing infinity.